dataset_api.md

```diff
@@ -0,0 +1,158 @@
+(rdflib.term.URIRef('http://example.com/subject-b'), rdflib.term.URIRef('http://example.com/predicate-b'), rdflib.term.Literal('Triple B'))
+
+# ============================================
+# Iterate over the entire Dataset returning quads
+# ============================================
+
+for quad in d.quads():
+    print(quad)
+# Output:
+(rdflib.term.URIRef('http://example.com/subject-a'), rdflib.term.URIRef('http://example.com/predicate-a'), rdflib.term.Literal('Triple A'), None)
+(rdflib.term.URIRef('http://example.com/subject-b'), rdflib.term.URIRef('http://example.com/predicate-b'), rdflib.term.Literal('Triple B'), rdflib.term.URIRef('http://example.com/graph-B'))
+
+# ============================================
+# Get the Default graph
+# ============================================
+
+dg = d.default_graph  # same as current default_context
+
+# ============================================
+# Iterate on triples in the Default Graph only
+# ============================================
+
+for triple in d.triples(graph="default"):
+    print(triple)
+# Output:
+(rdflib.term.URIRef('http://example.com/subject-a'), rdflib.term.URIRef('http://example.com/predicate-a'), rdflib.term.Literal('Triple A'))
+
+# ============================================
+# Access quads in Named Graphs only
+# ============================================
+
+for quad in d.quads(graph="named"):
+    print(quad)
+# Output:
+(rdflib.term.URIRef('http://example.com/subject-b'), rdflib.term.URIRef('http://example.com/predicate-b'), rdflib.term.Literal('Triple B'), rdflib.term.URIRef('http://example.com/graph-B'))
+
+# ============================================
+# Equivalent to iterating over graphs()
+# ============================================
+
+for ng_name, ng_object in d.graphs().items():
+    for quad in d.quads(graph=ng_name):
+        print(quad)
+# Output:
+(rdflib.term.URIRef('http://example.com/subject-b'), rdflib.term.URIRef('http://example.com/predicate-b'), rdflib.term.Literal('Triple B'), rdflib.term.URIRef('http://example.com/graph-B'))
+
+# ============================================
+# Access triples in the Default Graph and specified Named Graphs.
+# ============================================
+
+for triple in d.triples(graph=["default", URIRef("http://example.com/graph-B")]):
+    print(triple)
+# Output:
+(rdflib.term.URIRef('http://example.com/subject-a'), rdflib.term.URIRef('http://example.com/predicate-a'), rdflib.term.Literal('Triple A'))
+(rdflib.term.URIRef('http://example.com/subject-b'), rdflib.term.URIRef('http://example.com/predicate-b'), rdflib.term.Literal('Triple B'))
+
+# ============================================
+# Access quads in the Default Graph and specified Named Graphs.
+# ============================================
+
+for quad in d.quads(graph=["default", URIRef("http://example.com/graph-B")]):
```

Comment:
```python
for quad in (q for q in d.quads() if q[3] in (None, URIRef("http://example.com/graph-B"))): 
```
not much longer really.

Reply:
Yes I think this is the point to get a broader consensus on. The way I see it, if including the graph parameter:
Pros:
- can restrict "named", "default" enums to only be used in the graph= attribute, and not in the quads methods.
- can separate concerns a bit better, similar to how dataset clauses are used in SPARQL. E.g. set up an instance with graph= to restrict the scope to certain named graphs, then at runtime graphs can be passed in using quads
- provides a convenience/clean interface for what is a common pattern (for me at least!)
Cons:
- two ways to do the same thing, as you've pointed out.

Reply:
Maybe I'm too used to Jena where `Dataset` is used via `getDefaultModel` and `getNamedModel`, but I don't really see myself needing the new parameters 🤷‍♂️